scarb/src/core/registry/client/http.rs

```diff
@@ -146,12 +150,81 @@ impl<'c> RegistryClient for HttpRegistryClient<'c> {
     }
 
     async fn supports_publish(&self) -> Result<bool> {
-        // TODO(mkaput): Publishing to HTTP registries is not implemented yet.
-        Ok(false)
+        Ok(self.index_config.load().await?.upload.is_some())
     }
 
-    async fn publish(&self, _package: Package, _tarball: FileLockGuard) -> Result<()> {
-        todo!("Publishing to HTTP registries is not implemented yet.")
+    async fn publish(&self, package: Package, tarball: FileLockGuard) -> Result<RegistryUpload> {
+        let auth_token = env::var("SCARB_REGISTRY_AUTH_TOKEN").map_err(|_| {
+            anyhow!(
+                "missing authentication token. \
+            help: make sure SCARB_REGISTRY_AUTH_TOKEN environment variable is set"
+            )
+        })?;
+
+        let path = tarball.path();
+        ensure!(
+            Path::new(path).exists(),
+            "cannot upload package - file does not exist at path: {}",
+            path
+        );
+
+        let file = tarball.into_async().into_file();
+        let metadata = file.metadata().await?;
+        ensure!(
+            metadata.len() < 5 * 1024 * 1024,
+            "package cannot be larger than `5` MB: found `{}`",
+            &metadata.len() / 1024 / 1024
+        );
+
+        let index_config = self.index_config.load().await?;
+
+        let file_part = Part::stream(Body::from(file))
+            .file_name(format!("{}_{}", &package.id.name, &package.id.version));
+        let form = Form::new().part("file", file_part);
+
+        let response = self
+            .config
+            .online_http()?
+            .post(
+                index_config
+                    .upload
+                    .clone()
+                    .ok_or_else(|| anyhow!("failed to fetch registry upload url"))?,
+            )
+            .header(AUTHORIZATION, format!("Bearer {}", auth_token))
+            .multipart(form)
+            .timeout(Duration::from_secs(60))
+            .send()
+            .await?;
+
+        match response.status() {
+            StatusCode::UNAUTHORIZED => Err(RegistryUpload::Unauthorized)
+                .map_err(|_| anyhow!("invalid authentication token")),
+            StatusCode::FORBIDDEN => Err(RegistryUpload::CannotPublish)
+                .map_err(|_| anyhow!("missing upload permissions or not the package owner")),
+            StatusCode::BAD_REQUEST => Err(RegistryUpload::VersionExists)
+                .map_err(|_| anyhow!("package `{}` already exists", &package.id)),
```

Comment:
This seems problematic because there can be a mismatch between the actual error and what Scarb displays. For example, uploading package with 11 keywords using CURL will result in the following error:
```json
{"status":400,"error":"['Expected at most 10 keywords per package.']"}
```

But uploading the same package via `scarb publish` results a different, misleading error message:
```sh
error: package `access_control v0.3.5 ([..]/access_control/Scarb.toml)` already exists
```

Maybe we should consider @mkaput’s [suggestion](https://github.com/software-mansion/scarb/pull/1570/files#r1745749154) to use the `anyhow::Error` objects and propagate errors directly from the registry instead of producing our own.

Otherwise, even if we address this case, future errors returned by registry could still end up getting converted into generic `“file corrupted during upload.”`.

What do you think?

Reply:
Hmm, that might be a good reason to apply the suggestion. Indeed keywords checks were added in the meantime, and they utilise the same error code. 

I do wonder however, if we should reuse error codes for completely different potential errors 🤔 I guess registry should return different codes for different problems, right?

Reply:
> I do wonder however, if we should reuse error codes for completely different potential errors 🤔 I guess registry should return different codes for different problems, right?

I get where you’re coming from, but IIUC we're dealing with HTTP errors codes here, which are meant to capture wide range of errors under the same error code, and it's [common](https://stackoverflow.com/questions/3290182/which-status-code-should-i-use-for-failed-validations-or-invalid-duplicates) to use `400 Bad Request` for validation errors.

Even if consider using `422` as suggested in the thread I linked above, we'd still likely to encounter same error codes for different errors, including those generated internally by Django.

We might consider defining our own non-http error codes for our custom errors (~~starknet~~ scarbs spec 😄), but I'm not sure that's really necessary right now.

Reply:
Carrying over from similar thread with Marek.
Yeah, you're right. It's better to just pass through the errors, and depend on codes and messages from the registry. I have refactored the enum and adjusted the code https://github.com/software-mansion/scarb/pull/1570/commits/a9f383a41251ca3d139048aa2bab8640fefbf2c3